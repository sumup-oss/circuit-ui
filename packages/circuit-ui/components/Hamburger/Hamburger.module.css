.button {
  border: 0;
}

.skeleton {
  position: relative;
  display: flex;
  justify-content: center;
  transform: translateY(-1px);
}

.skeleton.s {
  width: var(--cui-icon-sizes-kilo);
  height: var(--cui-icon-sizes-kilo);
}

.skeleton.m {
  width: var(--cui-icon-sizes-mega);
  height: var(--cui-icon-sizes-mega);
}

.base {
  top: 50%;
  width: var(--hamburger-width);
}

.base,
.base::after,
.base::before {
  --hamburger-height: 2px;

  position: absolute;
  display: block;
  height: var(--hamburger-height);
  background-color: currentColor;
  border-radius: 1px;
  transition:
    width 0.2s ease-out 0.15s,
    opacity 0.1s ease-in,
    transform 0.3s cubic-bezier(0.55, 0.055, 0.675, 0.19);
}

.base::before,
.base::after {
  top: 0;
  content: "";
}

.base::before {
  width: calc(var(--hamburger-width) * 0.64);
  transform: translateY(calc(-1 * var(--hamburger-offset)));
}

.base::after {
  width: calc(var(--hamburger-width) * 0.82);
  transform: translateY(var(--hamburger-offset));
}

.base.s,
.base.s::after,
.base.s::before {
  --hamburger-width: calc(var(--cui-icon-sizes-kilo) - 2px);
  --hamburger-offset: 5px;
}

.base.m,
.base.m::after,
.base.m::before {
  --hamburger-width: calc(var(--cui-icon-sizes-mega) - 2px);
  --hamburger-offset: 7px;
}

[aria-pressed="true"] .base {
  transform: rotate(225deg);
}

[aria-pressed="true"] .base,
[aria-pressed="true"] .base::before,
[aria-pressed="true"] .base::after {
  width: var(--hamburger-width);
  transition:
    width 0.2s ease-out 0.15s,
    opacity 0.1s ease-out 0.15s,
    transform 0.3s cubic-bezier(0.215, 0.61, 0.355, 1) 0.15s;
}

[aria-pressed="true"] .base::before {
  opacity: 0;
  transform: translateY(0);
}

[aria-pressed="true"] .base::after {
  transform: translateY(0) rotate(-90deg);
}
