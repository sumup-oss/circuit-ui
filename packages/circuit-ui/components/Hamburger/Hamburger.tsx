/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { css } from '@emotion/react';

import styled, { StyleProps } from '../../styles/styled';
import { TrackingProps } from '../../hooks/useClickEvent';
import { IconButton, IconButtonProps } from '../IconButton/IconButton';
import { Skeleton } from '../Skeleton';
import { AccessibilityError } from '../../util/errors';

export type HamburgerRef = HTMLButtonElement;

export interface HamburgerProps
  extends Omit<IconButtonProps, 'ref' | 'children' | 'label' | 'type'> {
  /**
   * When active, the Hamburger transform into a close button.
   */
  isActive?: boolean;
  /**
   * Label for the 'active' state. Important for accessibility.
   */
  activeLabel: string;
  /**
   * Label for the 'inactive' state. Important for accessibility.
   */
  inactiveLabel: string;
  /**
   * @deprecated
   *
   * Use an `onClick` handler to dispatch user interaction events instead.
   */
  tracking?: TrackingProps;
  isLoading?: never;
  loadingLabel?: never;
}

const LAYER_HEIGHT = '2px';

const buttonStyles = ({ theme, size }: StyleProps & IconButtonProps) => css`
  border: 0;
  padding: ${size === 'giga' ? theme.spacings.kilo : theme.spacings.byte};
`;

const Button = styled(IconButton)<IconButtonProps>(buttonStyles);

type BoxProps = Required<Pick<HamburgerProps, 'size'>>;

const boxStyles = css`
  position: relative;
  display: flex;
  justify-content: center;
  transform: translateY(-1px);
`;

const boxSizeStyles = ({ theme, size }: StyleProps & BoxProps) => {
  const iconSizeMap = {
    giga: 'mega',
    kilo: 'kilo',
  } as const;
  const iconSize = iconSizeMap[size];

  return css`
    width: ${theme.iconSizes[iconSize]};
    height: ${theme.iconSizes[iconSize]};
  `;
};

const Box = styled(Skeleton)<BoxProps>(boxStyles, boxSizeStyles);

type LayerProps = Required<Pick<HamburgerProps, 'size' | 'isActive'>>;

const widthMap = {
  kilo: '14px',
  giga: '22px',
} as const;

const offsetMap = {
  kilo: '5px',
  giga: '7px',
} as const;

const layersBaseStyles = ({ size }: LayerProps) => {
  const width = widthMap[size];
  const offset = offsetMap[size];

  return css`
    top: 50%;
    width: ${width};

    &,
    &::after,
    &::before {
      background-color: currentColor;
      border-radius: 1px;
      display: block;
      height: ${LAYER_HEIGHT};
      position: absolute;
      transition: width 0.2s ease-out 0.15s, opacity 0.1s ease-in,
        transform 0.3s cubic-bezier(0.55, 0.055, 0.675, 0.19);
    }

    &::before,
    &::after {
      top: 0;
      content: '';
    }

    &::before {
      transform: translateY(-${offset});
      width: calc(${width} * 0.64);
    }

    &::after {
      transform: translateY(${offset});
      width: calc(${width} * 0.82);
    }
  `;
};

const layersActiveStyles = ({ isActive, size }: LayerProps) => {
  if (!isActive) {
    return null;
  }

  const width = widthMap[size];

  return css`
    transform: rotate(225deg);

    &,
    &::before,
    &::after {
      transition: width 0.2s ease-out 0.15s, opacity 0.1s ease-out 0.15s,
        transform 0.3s cubic-bezier(0.215, 0.61, 0.355, 1) 0.15s;
      width: ${width};
    }

    &::before {
      opacity: 0;
      transform: translateY(0);
    }

    &::after {
      transform: translateY(0) rotate(-90deg);
    }
  `;
};

const Layers = styled('span')<LayerProps>(layersBaseStyles, layersActiveStyles);

/**
 * A hamburger button for menus. Morphs into a close icon when active.
 */
export const Hamburger = ({
  isActive = false,
  activeLabel,
  inactiveLabel,
  size = 'giga',
  tracking,
  ...props
}: HamburgerProps): JSX.Element => {
  if (
    process.env.NODE_ENV !== 'production' &&
    process.env.NODE_ENV !== 'test'
  ) {
    if (!activeLabel) {
      throw new AccessibilityError(
        'Hamburger',
        'The `activeLabel` prop is missing.',
      );
    }
    if (!inactiveLabel) {
      throw new AccessibilityError(
        'Hamburger',
        'The `inactiveLabel` prop is missing.',
      );
    }
  }

  return (
    <Button
      {...props}
      size={size}
      label={isActive ? activeLabel : inactiveLabel}
      tracking={tracking ? { component: 'hamburger', ...tracking } : undefined}
      type="button"
    >
      <Box size={size}>
        <Layers isActive={isActive} size={size} />
      </Box>
    </Button>
  );
};
