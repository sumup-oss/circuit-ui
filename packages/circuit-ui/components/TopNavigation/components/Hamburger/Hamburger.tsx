/**
 * Copyright 2021, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { HTMLProps, MouseEvent, KeyboardEvent } from 'react';
import { css } from '@emotion/core';
import { Dispatch as TrackingProps } from '@sumup/collector';

import styled, { StyleProps } from '../../../../styles/styled';
import { focusVisible, hideVisually } from '../../../../styles/style-mixins';
import { useClickEvent } from '../../../../hooks/useClickEvent';

const HAMBURGER_WIDTH = '14px';
const LAYER_HEIGHT = '2px';

type OpenProps = Pick<HamburgerProps, 'isOpen'>;

const buttonBaseStyles = ({ theme }: StyleProps) => css`
  background: ${theme.colors.white};
  outline: none;
  border: none;
  border-right: ${theme.borderWidth.kilo} solid ${theme.colors.n300};
  border-bottom: ${theme.borderWidth.kilo} solid ${theme.colors.n300};
  margin: 0;
  padding: ${theme.spacings.kilo};
  transition: color ${theme.transitions.default},
    background-color ${theme.transitions.default},
    border-color ${theme.transitions.default};

  &:hover {
    cursor: pointer;
    background-color: ${theme.colors.n100};
  }

  ${focusVisible('inset')(theme)};

  ${theme.mq.mega} {
    display: none;
  }
`;

const buttonOpenStyles = ({ theme, isOpen }: StyleProps & OpenProps) =>
  isOpen &&
  css`
    background: ${theme.colors.black};
    border-right: ${theme.borderWidth.kilo} solid ${theme.colors.black};
    border-bottom: ${theme.borderWidth.kilo} solid ${theme.colors.black};

    &:hover {
      background-color: ${theme.colors.n900};
    }
  `;

const Button = styled.button<OpenProps>(buttonBaseStyles, buttonOpenStyles);

const boxStyles = ({ theme }: StyleProps) => css`
  position: relative;
  display: flex;
  justify-content: center;
  width: ${theme.iconSizes.mega};
  height: ${theme.iconSizes.mega};
`;

const Box = styled.span(boxStyles);

const layersBaseStyles = ({ theme }: StyleProps) => css`
  top: 50%;
  width: ${HAMBURGER_WIDTH};

  &,
  &::after,
  &::before {
    background-color: ${theme.colors.black};
    display: block;
    height: ${LAYER_HEIGHT};
    position: absolute;
    transition: width 0.2s ease-out 0.15s, opacity 0.1s ease-in,
      transform 0.3s cubic-bezier(0.55, 0.055, 0.675, 0.19);
  }

  &::before,
  &::after {
    top: 0;
    content: '';
  }

  &::before {
    transform: translateY(-4px);
    width: ${HAMBURGER_WIDTH};
  }

  &::after {
    transform: translateY(4px);
    width: ${HAMBURGER_WIDTH};
  }
`;

const layersOpenStyles = ({ theme, isOpen }: StyleProps & OpenProps) =>
  isOpen &&
  css`
    transform: rotate(225deg);

    &,
    &::before,
    &::after {
      background-color: ${theme.colors.white};
      border-radius: 1px;
      transition: width 0.2s ease-out, opacity 0.1s ease-out 0.15s,
        transform 0.3s cubic-bezier(0.215, 0.61, 0.355, 1) 0.15s;
      width: ${HAMBURGER_WIDTH};
    }

    &::before {
      opacity: 0;
      transform: translateY(0);
    }

    &::after {
      transform: translateY(0) rotate(-90deg);
    }
  `;

const Layers = styled('span')<OpenProps>(layersBaseStyles, layersOpenStyles);

export interface HamburgerProps extends HTMLProps<HTMLButtonElement> {
  /**
   * Function that's called when the hamburger is clicked.
   */
  onToggle?: (event: MouseEvent | KeyboardEvent) => void;
  /**
   * When open, the hamburger icon transform into a close icon.
   */
  isOpen?: boolean;
  /**
   * Label for the `open` state. Important for accessibility.
   */
  openLabel: string;
  /**
   * Label for the `closed` state. Important for accessibility.
   */
  closedLabel: string;
  /**
   * Additional data that is dispatched with the tracking event.
   */
  tracking?: TrackingProps;
}

export function Hamburger({
  isOpen = true,
  onToggle,
  openLabel = 'Close menu',
  closedLabel = 'Open menu',
  tracking = {},
  ...props
}: HamburgerProps): JSX.Element {
  const handleClick = useClickEvent(
    onToggle,
    tracking,
    `hamburger-${isOpen ? 'open' : 'closed'}`,
  );
  const label = isOpen ? openLabel : closedLabel;
  return (
    <Button
      {...props}
      isOpen={isOpen}
      onClick={handleClick}
      type="button"
      title={label}
    >
      <span css={hideVisually}>{label}</span>
      <Box>
        <Layers isOpen={isOpen} />
      </Box>
    </Button>
  );
}
