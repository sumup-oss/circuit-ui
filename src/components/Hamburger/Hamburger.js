/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { hideVisually } from 'polished';

const LAYER_HEIGHT = '1px';
const HAMBURGUER_WIDTH = '12px';

const buttonBaseStyles = ({ theme }) => css`
  label: hamburger;
  cursor: pointer;
  display: flex;
  justify-content: center;
  width: ${theme.iconSizes.giga};
  height: ${theme.iconSizes.giga};
  background: none;
  border: ${theme.borderWidth.kilo} solid ${theme.colors.n300};
  border-radius: 50%;
  position: relative;
`;

const lightButtonStyles = ({ theme, light }) =>
  light &&
  css`
    label: hamburger--light;
    border: ${theme.borderWidth.kilo} solid ${theme.colors.n700};
  `;

const layersBaseStyles = ({ theme }) => css`
  label: hamburger__layers;
  top: 50%;
  width: ${HAMBURGUER_WIDTH};

  &,
  &::after,
  &::before {
    background-color: ${theme.colors.n900};
    border-radius: ${theme.borderRadius.kilo};
    display: block;
    height: ${LAYER_HEIGHT};
    position: absolute;
    transition: width 0.2s ease-out 0.15s, opacity 0.1s ease-in,
      transform 0.3s cubic-bezier(0.55, 0.055, 0.675, 0.19);
  }

  &::before,
  &::after {
    top: 0;
    content: '';
  }

  &::before {
    transform: translateY(-${theme.spacings.bit});
    width: ${HAMBURGUER_WIDTH};
  }

  &::after {
    transform: translateY(${theme.spacings.bit});
    width: ${HAMBURGUER_WIDTH};
  }
`;

const layersBaseLightStyles = ({ theme, light }) =>
  light &&
  css`
    label: hamburger__layers--light;
    &,
    &::after,
    &::before {
      background-color: ${theme.colors.n100};
    }
  `;

const layersActiveStyles = ({ isActive }) =>
  isActive &&
  css`
    label: hamburger__layers--active;
    transform: rotate(225deg);

    &,
    &::before,
    &::after {
      transition: width 0.2s ease-out, opacity 0.1s ease-out 0.15s,
        transform 0.3s cubic-bezier(0.215, 0.61, 0.355, 1) 0.15s;
      width: ${HAMBURGUER_WIDTH};
    }

    &::before {
      opacity: 0;
      transform: translateY(0);
    }

    &::after {
      transform: translateY(0) rotate(-90deg);
    }
  `;

const labelBaseStyles = () => css`
  label: hamburger__label;
  ${hideVisually()};
`;

const HamburgerButton = styled('button')`
  ${buttonBaseStyles}
  ${lightButtonStyles};
`;

const HamburgerLayers = styled('span')`
  ${layersBaseStyles}
  ${layersActiveStyles}
  ${layersBaseLightStyles};
`;
const HamburgerLabel = styled('span')`
  ${labelBaseStyles};
`;

/**
 * A hamburger button for menus. Morphs into a close icon when active.
 */
const Hamburger = ({
  onClick,
  isActive,
  labelActive,
  labelInActive,
  light,
  ...rest
}) => (
  <HamburgerButton {...rest} onClick={onClick} light={light}>
    <HamburgerLayers isActive={isActive} light={light} />
    <HamburgerLabel>{isActive ? labelActive : labelInActive}</HamburgerLabel>
  </HamburgerButton>
);

Hamburger.propTypes = {
  /**
   * Function that is called with the event when the hamburger is clicked.
   */
  onClick: PropTypes.func,
  /**
   * A consice description of the example prop.
   */
  isActive: PropTypes.bool,
  /**
   * Label for the 'active' state. Important for accessibility.
   */
  labelActive: PropTypes.string,
  /**
   * Label for the 'inactive' state. Important for accessibility.
   */
  labelInActive: PropTypes.string,
  /**
   * A Boolean to select the Light or Dark (default) version of the Hamburger.
   */
  light: PropTypes.bool,
  className: PropTypes.string
};

Hamburger.defaultProps = {
  onClick: () => {},
  isActive: false,
  labelActive: 'Close menu',
  labelInActive: 'Open menu'
};

/**
 * @component
 */
export default Hamburger;
