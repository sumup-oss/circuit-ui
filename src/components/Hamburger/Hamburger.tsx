/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FC, HTMLProps } from 'react';
import { css } from '@emotion/core';

import styled, { StyleProps } from '../../styles/styled';
import { IconButton, IconButtonProps } from '../IconButton/IconButton';

export type HamburgerRef = HTMLButtonElement;

export interface HamburgerProps
  extends Omit<IconButtonProps, 'ref' | 'children' | 'label'> {
  /**
   * When active, the Hamburger transform into a close button.
   */
  isActive?: boolean;
  /**
   * Label for the 'active' state. Important for accessibility.
   */
  labelActive: string;
  /**
   * Label for the 'inactive' state. Important for accessibility.
   */
  labelInActive: string;
  'data-testid'?: string;
}

const LAYER_HEIGHT = '2px';
const HAMBURGER_WIDTH = '14px';

const buttonStyles = () => css`
  label: hamburger;
  border: 0;
`;

const Button = styled(IconButton)<IconButtonProps>(buttonStyles);

const boxStyles = ({ theme }: StyleProps) => css`
  label: hamburger__box;
  position: relative;
  display: flex;
  justify-content: center;
  width: ${theme.iconSizes.mega};
  height: ${theme.iconSizes.mega};
  transform: translateY(-1px);
`;

const Box = styled.span<HTMLProps<HTMLSpanElement>>(boxStyles);

const layersBaseStyles = ({ theme }: StyleProps) => css`
  label: hamburger__layers;
  top: 50%;
  width: ${HAMBURGER_WIDTH};

  &,
  &::after,
  &::before {
    background-color: currentColor;
    border-radius: ${theme.borderRadius.kilo};
    display: block;
    height: ${LAYER_HEIGHT};
    position: absolute;
    transition: width 0.2s ease-out 0.15s, opacity 0.1s ease-in,
      transform 0.3s cubic-bezier(0.55, 0.055, 0.675, 0.19);
  }

  &::before,
  &::after {
    top: 0;
    content: '';
  }

  &::before {
    transform: translateY(-5px);
    width: ${HAMBURGER_WIDTH};
  }

  &::after {
    transform: translateY(5px);
    width: ${HAMBURGER_WIDTH};
  }
`;

const layersActiveStyles = ({ isActive }: { isActive?: boolean }) =>
  isActive &&
  css`
    label: hamburger__layers--active;
    transform: rotate(225deg);

    &,
    &::before,
    &::after {
      transition: width 0.2s ease-out, opacity 0.1s ease-out 0.15s,
        transform 0.3s cubic-bezier(0.215, 0.61, 0.355, 1) 0.15s;
      width: ${HAMBURGER_WIDTH};
    }

    &::before {
      opacity: 0;
      transform: translateY(0);
    }

    &::after {
      transform: translateY(0) rotate(-90deg);
    }
  `;

const Layers = styled('span')<{ isActive?: boolean }>(
  layersBaseStyles,
  layersActiveStyles
);

/**
 * A hamburger button for menus. Morphs into a close icon when active.
 */
export const Hamburger: FC<HamburgerProps> = ({
  isActive,
  labelActive = 'Close menu',
  labelInActive = 'Open menu',
  ...props
}: HamburgerProps) => (
  <Button label={isActive ? labelActive : labelInActive} size="kilo" {...props}>
    <Box>
      <Layers isActive={isActive} />
    </Box>
  </Button>
);
