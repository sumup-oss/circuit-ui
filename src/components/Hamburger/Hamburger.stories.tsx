/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';

import { Hamburger, HamburgerProps } from './Hamburger';

export default {
  title: 'Components/Hamburger',
  component: Hamburger,
};

export const Base = (args: HamburgerProps) => {
  const [active, setActive] = useState(false);
  const handleClick = () => {
    setActive((prev) => !prev);
  };
  return <Hamburger isActive={active} onClick={handleClick} {...args} />;
};

Base.args = {
  labelActive: 'Close menu',
  labelInActive: 'Open menu',
};
