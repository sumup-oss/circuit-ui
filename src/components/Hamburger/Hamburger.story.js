/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';
import { text, boolean } from '@storybook/addon-knobs/react';

import Hamburger from './Hamburger';

export default {
  title: 'Components|Hamburger',
  component: Hamburger,
  parameters: {
    jest: ['Hamburger']
  }
};

const HamburderWithState = () => {
  const [active, setActive] = useState(false);
  const handleClick = () => {
    setActive(prev => !prev);
  };
  return <Hamburger isActive={active} onClick={handleClick} />;
};

export const base = () => (
  <HamburderWithState
    labelActive={text('Label active', 'Close menu')}
    labelInActive={text('Label inactive', 'Open menu')}
    light={boolean('light', false)}
  />
);
