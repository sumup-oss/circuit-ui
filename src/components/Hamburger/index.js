import Hamburger from './Hamburger';

export default Hamburger;
