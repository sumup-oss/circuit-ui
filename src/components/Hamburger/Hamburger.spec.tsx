/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import {
  create,
  renderToHtml,
  axe,
  render,
  act,
  userEvent,
  RenderFn
} from '../../util/test-utils';

import { Hamburger, HamburgerProps } from './Hamburger';

describe('Hamburger', () => {
  function renderHamburger(renderFn: RenderFn, props: HamburgerProps) {
    return renderFn(<Hamburger {...props} />);
  }

  const baseProps = {
    labelActive: 'Close menu',
    labelInActive: 'Open menu'
  };

  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = renderHamburger(create, baseProps);
    expect(actual).toMatchSnapshot();
  });

  it('should render with active styles when passed the isActive prop', () => {
    const actual = renderHamburger(create, { ...baseProps, isActive: true });
    expect(actual).toMatchSnapshot();
  });

  /**
   * Logic tests.
   */
  it('should call the onClick prop when clicked', () => {
    const onClick = jest.fn();
    const { getByTestId } = renderHamburger(render, {
      ...baseProps,
      onClick,
      'data-testid': 'hamburger'
    });

    act(() => {
      userEvent.click(getByTestId('hamburger'));
    });

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderHamburger(renderToHtml, baseProps);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
