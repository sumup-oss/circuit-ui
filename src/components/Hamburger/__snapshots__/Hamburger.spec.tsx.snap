// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hamburger should render with active styles when passed the isActive prop 1`] = `
.circuit-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: auto;
  cursor: pointer;
  font-size: 16px;
  line-height: 24px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  border-radius: 5px;
  border-width: 1px;
  border-style: solid;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
  background-color: #FFFFFF;
  border-color: #9DA7B1;
  color: #323E49;
  padding: 8px calc(24px - 1px);
  border: 0;
}

.circuit-3:focus {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-3:focus::-moz-focus-inner {
  border: 0;
}

.circuit-3:disabled,
.circuit-3[disabled] {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
}

.circuit-3:hover {
  background-color: #FAFBFC;
  border-color: #5C656F;
}

.circuit-3:active {
  background-color: #EEF0F2;
  border-color: #5C656F;
}

.circuit-1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
}

.circuit-2 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.circuit-0 {
  top: 50%;
  width: 14px;
  -webkit-transform: rotate(225deg);
  -ms-transform: rotate(225deg);
  transform: rotate(225deg);
}

.circuit-0,
.circuit-0::after,
.circuit-0::before {
  background-color: currentColor;
  border-radius: 1px;
  display: block;
  height: 2px;
  position: absolute;
  -webkit-transition: width 0.2s ease-out 0.15s,opacity 0.1s ease-in, -webkit-transform 0.3s cubic-bezier(0.55,0.055,0.675,0.19);
  -webkit-transition: width 0.2s ease-out 0.15s,opacity 0.1s ease-in, transform 0.3s cubic-bezier(0.55,0.055,0.675,0.19);
  transition: width 0.2s ease-out 0.15s,opacity 0.1s ease-in, transform 0.3s cubic-bezier(0.55,0.055,0.675,0.19);
}

.circuit-0::before,
.circuit-0::after {
  top: 0;
  content: '';
}

.circuit-0::before {
  -webkit-transform: translateY(-5px);
  -ms-transform: translateY(-5px);
  transform: translateY(-5px);
  width: 14px;
}

.circuit-0::after {
  -webkit-transform: translateY(5px);
  -ms-transform: translateY(5px);
  transform: translateY(5px);
  width: 14px;
}

.circuit-0,
.circuit-0::before,
.circuit-0::after {
  -webkit-transition: width 0.2s ease-out,opacity 0.1s ease-out 0.15s, -webkit-transform 0.3s cubic-bezier(0.215,0.61,0.355,1) 0.15s;
  -webkit-transition: width 0.2s ease-out,opacity 0.1s ease-out 0.15s, transform 0.3s cubic-bezier(0.215,0.61,0.355,1) 0.15s;
  transition: width 0.2s ease-out,opacity 0.1s ease-out 0.15s, transform 0.3s cubic-bezier(0.215,0.61,0.355,1) 0.15s;
  width: 14px;
}

.circuit-0::before {
  opacity: 0;
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.circuit-0::after {
  -webkit-transform: translateY(0) rotate(-90deg);
  -ms-transform: translateY(0) rotate(-90deg);
  transform: translateY(0) rotate(-90deg);
}

<button
  class="circuit-3"
  title="Close menu"
>
  <span
    class="circuit-1"
    role="presentation"
  >
    <span
      class="circuit-0"
    />
  </span>
  <span
    class="circuit-2"
  >
    Close menu
  </span>
</button>
`;

exports[`Hamburger should render with default styles 1`] = `
.circuit-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: auto;
  cursor: pointer;
  font-size: 16px;
  line-height: 24px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  border-radius: 5px;
  border-width: 1px;
  border-style: solid;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
  background-color: #FFFFFF;
  border-color: #9DA7B1;
  color: #323E49;
  padding: 8px calc(24px - 1px);
  border: 0;
}

.circuit-3:focus {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-3:focus::-moz-focus-inner {
  border: 0;
}

.circuit-3:disabled,
.circuit-3[disabled] {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
}

.circuit-3:hover {
  background-color: #FAFBFC;
  border-color: #5C656F;
}

.circuit-3:active {
  background-color: #EEF0F2;
  border-color: #5C656F;
}

.circuit-1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  -webkit-transform: translateY(-1px);
  -ms-transform: translateY(-1px);
  transform: translateY(-1px);
}

.circuit-0 {
  top: 50%;
  width: 14px;
}

.circuit-0,
.circuit-0::after,
.circuit-0::before {
  background-color: currentColor;
  border-radius: 1px;
  display: block;
  height: 2px;
  position: absolute;
  -webkit-transition: width 0.2s ease-out 0.15s,opacity 0.1s ease-in, -webkit-transform 0.3s cubic-bezier(0.55,0.055,0.675,0.19);
  -webkit-transition: width 0.2s ease-out 0.15s,opacity 0.1s ease-in, transform 0.3s cubic-bezier(0.55,0.055,0.675,0.19);
  transition: width 0.2s ease-out 0.15s,opacity 0.1s ease-in, transform 0.3s cubic-bezier(0.55,0.055,0.675,0.19);
}

.circuit-0::before,
.circuit-0::after {
  top: 0;
  content: '';
}

.circuit-0::before {
  -webkit-transform: translateY(-5px);
  -ms-transform: translateY(-5px);
  transform: translateY(-5px);
  width: 14px;
}

.circuit-0::after {
  -webkit-transform: translateY(5px);
  -ms-transform: translateY(5px);
  transform: translateY(5px);
  width: 14px;
}

.circuit-2 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<button
  class="circuit-3"
  title="Open menu"
>
  <span
    class="circuit-1"
    role="presentation"
  >
    <span
      class="circuit-0"
    />
  </span>
  <span
    class="circuit-2"
  >
    Open menu
  </span>
</button>
`;
